drivers/usb/usbstor/pdo.c

```diff
@@ -437,37 +437,60 @@ USBSTOR_PdoHandleQueryInstanceId(
 {
     PPDO_DEVICE_EXTENSION PDODeviceExtension;
     PFDO_DEVICE_EXTENSION FDODeviceExtension;
-    WCHAR Buffer[100];
-    ULONG Length;
+    PUSB_STRING_DESCRIPTOR Descriptor;
+    ULONG CharCount;
     LPWSTR InstanceId;
+    NTSTATUS Status;
 
-    PDODeviceExtension = (PPDO_DEVICE_EXTENSION)DeviceObject->DeviceExtension;
-    FDODeviceExtension = (PFDO_DEVICE_EXTENSION)PDODeviceExtension->LowerDeviceObject->DeviceExtension;
+    PDODeviceExtension = DeviceObject->DeviceExtension;
+    FDODeviceExtension = PDODeviceExtension->LowerDeviceObject->DeviceExtension;
 
-    // format instance id
-    if (FDODeviceExtension->SerialNumber)
+    Descriptor = FDODeviceExtension->SerialNumber;
+    if (Descriptor && (Descriptor->bLength >= sizeof(USB_COMMON_DESCRIPTOR) + sizeof(WCHAR)))
     {
-        // using serial number from device
-        swprintf(Buffer, L"%s&%c", FDODeviceExtension->SerialNumber->bString, PDODeviceExtension->LUN);
+        /* Format the serial number descriptor only if supported by the device */
+        CharCount = (Descriptor->bLength - sizeof(USB_COMMON_DESCRIPTOR)) / sizeof(WCHAR) +
+                    (sizeof("&") - 1) +
+                    (sizeof("9") - 1) + // LUN: 1 char
```

Comment:
```suggestion
                    (sizeof("FF") - 1) + // LUN: 2 char
```
The correct one is a row of 2 characters, because the value of FF fits into an [UCHAR](https://git.reactos.org/?p=reactos.git;a=blob;f=drivers/usb/usbstor/usbstor.h;hb=0bce79a50b52bce96b54ab334bb0bf7fe742feb8#l155) type.

Reply:
Yes, but from what I saw [here](https://git.reactos.org/?p=reactos.git;a=blob;f=drivers/usb/usbstor/misc.c;hb=0bce79a50b52bce96b54ab334bb0bf7fe742feb8#l192), the value is limited by [MAX_LUN](https://git.reactos.org/?p=reactos.git;a=blob;f=drivers/usb/usbstor/usbstor.h;hb=0bce79a50b52bce96b54ab334bb0bf7fe742feb8#l61). I can increase it to 2 chars if preferable though.

Reply:
It would be nice to have a comment to reference the MAX_LUN define.